ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+    let url = URL(string: "\(apiBaseURL)languages?device=\(deviceType)")!
+    let userData = completionBlock.map { ["completionBlock": $0] }
+
+    let request = HTTPDownloadRequest(url: url, downloadType: .downloadCachedData, userInfo: userData ?? [:])
+    currentRequest = request
+    sharedQueue.addRequest(request)
+    sharedQueue.run()
+  }
+
+  /// Asynchronously fetches the .js file for the keyboard with given IDs.
+  /// See `NSNotification+Notifications` for notification on success/failiure.
+  /// - Parameters:
+  ///   - isUpdate: Keep the keyboard files on failure
+  public func downloadKeyboard(withID keyboardID: String, languageID: String, isUpdate: Bool) {
+    let kbInfo = [
+      Key.keyboardId: keyboardID,
+      Key.languageId: languageID
+    ]
+
+    guard downloadQueue == nil else {
+      // Download queue is active.
+      return
+    }
+
+    guard reachability.currentReachabilityStatus() != NotReachable else {
+      let error = NSError(domain: "Keyman", code: 0,
+                          userInfo: [NSLocalizedDescriptionKey: "No internet connection"])
+      downloadFailed(forKeyboard: kbInfo, error: error)
+      return
+    }
+
+    guard let keyboardsInfo = keyboardsInfo else {
+      let message = "Keyboard info has not yet been fetched. Call fetchKeyboardsList() first."
+      let error = NSError(domain: "Keyman", code: 0, userInfo: [NSLocalizedDescriptionKey: message])
+      downloadFailed(forKeyboard: kbInfo, error: error)
+      return
+    }
+
+    let kbKey = "\(languageID)_\(keyboardID)"
+    guard let keyboardDict = keyboardsDictionary[kbKey] else {
+      let error = NSError(domain: "Keyman", code: 0,
+                          userInfo: [NSLocalizedDescriptionKey: "Keyboard not found with key: \(kbKey)"])
+      downloadFailed(forKeyboard: kbInfo, error: error)
+      return
+    }
+
+    let urlStr = keyboardsInfo[keyboardID]![Key.keyboardURI]!
+    let keyboardURL = URL(string: urlStr)!
+
+    let kbFontStr = keyboardDict[Key.font] ?? ""
+    let kbOskFontStr = keyboardDict[Key.oskFont] ?? ""
+    var kbFontURLs: [URL] = []
+    do {
+      kbFontURLs += try keyboardFontURLs(jsonString: kbFontStr)
+      kbFontURLs += try keyboardFontURLs(jsonString: kbOskFontStr)
+    } catch {
+      downloadFailed(forKeyboard: kbInfo, error: error as NSError)
+      return
+    }
+
+    // TODO: Better typing
+    downloadQueue = HTTPDownloader(self)
+    let commonUserData: [String: Any] = [
+      Key.keyboardInfo: kbInfo,
+      Key.update: isUpdate ? 1 : 0
+    ]
+    downloadQueue!.userInfo = commonUserData
+
+    let kbVersion = keyboardDict[Key.keyboardVersion]
+    let keyboardPath = self.keyboardPath(forFilename: keyboardURL.lastPathComponent,
+                                         keyboardVersion: kbVersion)
+
+    var request = HTTPDownloadRequest(url: keyboardURL, userInfo: commonUserData)
+    request.destinationFile = keyboardPath?.path
```

Comment:
Not `keyboardPath`, but `keyboardPath?.path`?  Why the change in structure compared to the original?  (Also affects the similar line in the loop immediately below it.)

Reply:
I changed a lot of the functions that locate files to return `URL` instead of `String` and `URL.path` gives the path as a `String`.